At this step all of you should have a fasta file for each 1-to-1 orthologs.

Since the specie tree inferred directly by Orthofinder was incorrect (and usually in phylogenomic no one trust in the raw output of Orthofinder), now we want to  try to make a better estimation of the phylogenetic relathionships between the 8 species under consideration, starting with a...

# Multiple Sequence Alignment

After we have get our strictly orthologs sequences, the next step is to infer the orthology relathionships between **each site** of each sequence. In other terms, the differences among the nucleotides/amminoacide should result **only from changes that convey descent information**.This process is called **Multiple Sequence Alignment (MSA)** and is often followed by an additional step of detection and exclusion of those alignment regions whose homology we are uncertain of. Due to the large amount of data which we process nowadays, this step is often overlooked and as a result it's quite easy to find misaligned loci in modern datasets. Moreover, from the bioinformatic poin of view, there is a lot under the hood of the aligners (see [here](https://www.sciencedirect.com/science/article/pii/S0888754317300551) for a nice review on underlying computational methods).


In this tutorial, we will use the most popular tools for multiple sequence alignment:

  * MAFFT (Original publication: [Katoh et al., 2002](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC135756/))
  * T-Coffee (Original publication: [Notredame et al., 2000](https://www.sciencedirect.com/science/article/abs/pii/S0022283600940427))
  * And subsequently filter the alignments using  the old but gold:
  * Gblocks ([Talavera & Castresana, 2007](https://academic.oup.com/mbe/article/17/4/540/1127654))
 
 Let's organize our working directories:
 
 ```
mkdir -p Analyses/1-to-1_Alignments/{mafft,t_coffee,gblock,mafft_dash}
 ```
 ### 1. Mafft (an evergreen)
 
 Print the help of mafft to get some few, but usefull informations:
 
 ```
 mafft --help
 ------------------------------------------------------------------------------
  MAFFT v7.475 (2020/Nov/23)
  https://mafft.cbrc.jp/alignment/software/
  MBE 30:772-780 (2013), NAR 30:3059-3066 (2002)
------------------------------------------------------------------------------
High speed:
  % mafft in > out
  % mafft --retree 1 in > out (fast)

High accuracy (for <~200 sequences x <~2,000 aa/nt):
  % mafft --maxiterate 1000 --localpair  in > out (% linsi in > out is also ok)
  % mafft --maxiterate 1000 --genafpair  in > out (% einsi in > out)
  % mafft --maxiterate 1000 --globalpair in > out (% ginsi in > out)

If unsure which option to use:
  % mafft --auto in > out

--op # :         Gap opening penalty, default: 1.53
--ep # :         Offset (works like gap extension penalty), default: 0.0
--maxiterate # : Maximum number of iterative refinement, default: 0
--clustalout :   Output: clustal format, default: fasta
--reorder :      Outorder: aligned, default: input order
--quiet :        Do not report progress
--thread # :     Number of threads (if unsure, --thread -1)
--dash :         Add structural information (Rozewicki et al, submitted)
 ```
 
For more informations about the alghoritms see the [mafft website](https://mafft.cbrc.jp/alignment/software/algorithms/algorithms.html). Just few notes:

 * **E-INS-i** strategy is designed to work for proteins which have several conserved motifs embedded in long unalignable regions (*e.g.* fast-evolving multi-domains proteins, Fig. 1A).
 * **L-INS-i** is suitable for single-domain proteins with unalignable regions (Fig. 1B).
 * **G-INS-i** is designed to maximase alignment accuracy between full alignable regions (*e.g* domains alignments, Fig. 1C).

![Examples](https://github.com/for-giobbe/phy/blob/master/2021/Images/Mafft-1.png)

In our case we don't know the architecture of our proteins, eve if we can confidentially exclude the case 1C. So we can take advantage of the ```--auto``` option.
Instead, in  other scenarios we should now which algorithm is more suitable for our data, for example when we want to infer the evolutionary history of a single domain (but take care about the number of sequences you want to align, these alghoritms can take forever!). Moreover we will add the ```--anysymbol``` option so mafft won't get stucked if it will find a IUPAC ambiguity code.

Let's align our 1-to-1 Orthologs and move them to the proper directory:

```
for i in Analyses/My_Orthology.Inference/Results_Apr09/Single_Copy_Orthologue_Sequences/*.fa; do \
 mafft --auto --anysymbol "$i" > "$i".mafft;
done;

mv Analyses/My_Orthology.Inference/Results_Apr09/Single_Copy_Orthologue_Sequences/OG00000*.mafft Analyses/1-to-1_Alignments/mafft/
```

### 2. T-Coffee

T-Coffe is a popular aligner with an incredibly nice server which I strongly encourage to explore. But today we are focusing on it's variant M-COFFEE: the idea behind M-COFFE is to combine many of the more popular aligners (including MAFFT). If we type ```t_coffee``` and scroll up a bit we can check wether the other aligners have been correctly installed (they should be, thanks to the magic behind conda installing). We should spot easily the popular ones as Probcons, Muscle, Clustal and so on. We can give it a try using the default combination of aligners (Mkalign, Muscle & MAFFT) by typing:

```
for i in Analyses/My_Orthology.Inference/Results_Apr09/Single_Copy_Orthologue_Sequences/*.fa; do \
 t_coffee -seq "$i" -output fasta -mode fmcoffee;
done;

mv OG* Analyses/1-to-1_Alignments/t_coffee/
```
where:

 * ```-mode fmcoffee``` specify the mode of t_coffee
 * ```-output fasta``` specify the output format (by default is in clustal, a quite problematic format).
 
We should also take a look at the outputs:

 * the **.aln** is the alignment file in the clustal format.
 * the **.dnd** file is the guide tree generated in the process.
 
 As you can see M-Coffee is combining multiple aligners into one and evaluating the uncertainty between their outcome.
 
 ### 3. Structural alignment: [Mafft-DASH](https://pubmed.ncbi.nlm.nih.gov/31062021/)

This method is able to incorporate structural alignments with the MAFFT tool. In brief, it queryes a web-based Database of Aligned Structural Homologs (DASH) with input sequences. Homologs with structural information are added to the alignment, guiding it. As you can imagine, this method is very suitable for domain alignments (indeed by default it uses the G-INS-i strategy), but it can also be used together with L-INS-i, as we are going to do.

**NOTE:** This method is **extremely** slow and can only be used when we have few proteins!

```
for i in Analyses/My_Orthology.Inference/Results_Apr09/Single_Copy_Orthologue_Sequences/*.fa; do \
 mafft --dash --localpair --originalseqonly --anysymbol "$i"  > "$i".mafftdash;
done;

mv Analyses/My_Orthology.Inference/Results_Apr09/Single_Copy_Orthologue_Sequences/*.mafftdash Analyses/1-to-1_Alignments/mafft_dash/
```

where:

 * ```--dash``` specify to run mafft in DASH mode.
 * ```--localpair``` specify to use the L-INS-i strategy instead of the default G-INS-i (probably not suitable for our data).
 * ```--anysymbol``` as before.


### Which is the best alignment?

The best alignment can be choosen in two main ways: by eye, if we have enough time and *"sensitivity"* (*i.e.* experience), or using specific tools (M-COFFEE function of T-COFFEE is one of them, but see also [TrimAl](http://trimal.cgenomics.org/)). However, usually we choose the aligner *a priori* based on the analyses that we want to carry on, on the size of the dataset ecc... For example, if we want to estimate the selection regimes between different lineages we should prefer the more precise aligner, regardless of the computational time (in human terms). For phylogenetic inference usually are preffered faster aligners since is more robust to errors (especially if we are using hundreds or thousands of 1 to 1 loci which should be highly conserved).

Now open all alignments for some of the OGs and take some time to look at them.

 * Can you identify more conserved regions?
 * Where are usually placed unalignable regions?

---

## Alignment filtering:

As previously said, the quality of multiple sequence alignments plays an important role in the accuracy of phylogenetic inference and even more in other analyses. It has been shown that removing certain positions (generally the ones which show an ambiguous homology and/or are highly variable) can improve the overall performance of many phylogenetic reconstruction methods. While dealing modern phylogenetic dataset, which consists of hundred to thousands of alignments is not possible to have a manual curation for each one, and thus it is necessary to automatically remove alignment errors.
<br/><br/>
In this tutorial we will use Gblocks: this software will select blocks of conserved sites, which can be defined with many custom parameters, described in the [manual](https://home.cc.umanitoba.ca/~psgendb/doc/Castresana/Gblocks_documentation.html)

We will use the susequent parameters and leave the rest to the default:

```
-t= p (protein)
-b5 Allowed Gap Positions
```

>Be cautios when using automatic filtering methods, some times they can overprune your alignments giving **worst** results (see [this](https://academic.oup.com/sysbio/article/64/5/778/1685763) publication). Generally is a good practice to try different software with different parameters for at least some of the alignments and take a look at the results by eye.

Now run gblocks on our alignments with a simple for loop (we will also redirect stdout and stderr to a log file):

```
for i in $( find ./Analyses/1-to-1_Alignments/ -type f \( -iname \*.mafft -o -iname \*.mafftdash -o -iname \*aln \)); do \
 Gblocks "$i" -t=p -b5=a 2>&1 | tee "$i".gb.log;
done
```

The analyses will generate two kind of files:

 * **-gb:** filtered fasta files.
 * **-gb.html:** for a nice visualization on which position got removed.

Now, we can roughly summarize results of gblock with a simple bash script using the log files:

```
for i in $( ls Analyses/1-to-1_Alignments/ | grep "mafft\|mafft_dash\|t_coffee"); do
 varResults=$( cat ./Analyses/1-to-1_Alignments/"$i"/*gb.log | grep "Gblocks alignment" | cut -d" " -f6 | sed 's/(//g' | awk '{ total += $1 } END { print total/NR }');
 echo -e "$i\t$varResults";
done

mafft	64.0612
mafft_dash	64.2041
t_cofee	64.6531
```

If we want to increase the number of selected blocks we can decrease the ```-b4``` parameter (Minimum Length Of A Block) from deafult value of 10 to 5 and increase the ```-b3``` flag (Maximum Number Of Contiguous Nonconserved Positions) from the default 8 to 10. We need also to specify the ```-e``` option (Generic File Extension) so previously results will not be overwritten:

```
for i in $( find ./Analyses/1-to-1_Alignments/ -type f \( -iname \*.mafft -o -iname \*.mafftdash -o -iname \*aln \)); do
 Gblocks "$i" -t=p -b4=8 -b3=10 -e=-gb2 -b5=a 2>&1 | tee "$i".gb-2.log;
done
```

Now, let's summarize again our new results:

```
for i in $( ls Analyses/1-to-1_Alignments/ | grep "mafft\|mafft_dash\|t_coffee"); do
 varResults=$( cat ./Analyses/1-to-1_Alignments/"$i"/*gb-2.log | grep "Gblocks alignment" | cut -d" " -f6 | sed 's/(//g' | awk '{ total += $1 } END { print total/NR }');
 echo -e "$i\t$varResults";
done

mafft	66.0408
mafft_dash	66.0816
t_cofee	66.2653
```
As we can see both set of parameters roughly conserved the same number of positions with the slightly better alignments of t_cofee (only in terms of positions retained by gblock!).
Now we can have a quick view of some of our cleaned alignments. For this purpose the html files are very usefull:

 * Which positions are usually discared in almost all alignments?
 * Try to find t_coffee alignments most affected by gblock analyses (using a for loop on the **log** files) and take a look at the corresponding html files.


For next steps we can use the t_coffee aligments filtered with default options of Gblock (*rationale*: t_coffee alignments have more conserved regions and since results haven't changed much between different set of parameters - at least with a quick overview - its easier to just use the ones obtained with default gblock parameters). Let's move these files in the proper ```Analyses/1-to-1_Alignments/gblock``` folder (being organized in the management of files/folders is **essential** when we are analyzing a lot of data):

```
mv -iv Analyses/1-to-1_Alignments/t_coffee/*-gb Analyses/1-to-1_Alignments/gblock/
```

If you have missed these data you can find them [here](https://github.com/for-giobbe/phy/blob/master/2021/Analyses/Precomputed_1-to-1_Alignments.tar.gz)

---

Many other alternatives are possible, and all this tools can be extremely helpfull in removing noise and ameliorating certain characteristic of phylogenetic datasets which can affect subsequent inferences, such as substitution saturation & compositional heterogeneity. Here are the most popular:

 * **Aliscore** ([Kück et al., 2014](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-294))
 * **BMGE** ([Criscuolo & Gibaldo](https://bmcecolevol.biomedcentral.com/articles/10.1186/1471-2148-10-210))
 * **HMMCleaner** (very interesting, it uses HMM profiles to identify primary sequence errors, see their pubblication: [Di Franco et al., 2019](https://bmcecolevol.biomedcentral.com/articles/10.1186/s12862-019-1350-2))
 * **Prequal** (very interesting 2x since it masks non-homologous regions **BEFORE** alignment step, see [Whelan et al., 2018](https://pubmed.ncbi.nlm.nih.gov/29868763/))
 * **TrimAl** (an evergreen, like Gblock; [http://trimal.cgenomics.org/])

Other approaches for filtering are possible and useful when dealing with a large number of loci. Positions, genes and entire alignment can be filtered out on several crietria, including:

 * length
 * occupancy
 * dNdS
...

---

## further reading: 

[very interesting paper on multiple sequence alignment uncertainty pt.1 ](https://academic.oup.com/gbe/article/7/8/2102/556628).

[very interesting paper on multiple sequence alignment uncertainty pt.2](https://science.sciencemag.org/content/319/5862/473?casa_token=t07ptffISm4AAAAA:j5l4US_y_GHOMduYw6R-MhyM7YUpa__08lw45l455DAU3tGFNKYlV40ZH0Si5w48Xl1gTEqsocLVvaE).
