# GENE FAMILIES EVOLUTIONARY ANALYSES

## INTRO: 

Evolutionary studies about gene families are becoming very frequent in recent years. For example, in almost all genome paper is present a gene families expansion/contractions analyses. Indeed, has been observed how enormous amount of changes in families size can take place even between closely related organisms. There is much interest in these changes, as even the gain or loss of single genes have been implicated in adaptive divergence between species. In addition, large contractions and expansions of gene families are generally attributed to natural selection.

One of the most popular software is **[CAFE](https://academic.oup.com/bioinformatics/article/22/10/1269/237347)**, even if other apporaches based on gene and species tree exist. Its purpose is to analyze changes in gene family size in a way that accounts for phylogenetic history and provides a statistical foundation for evolutionary inferences. The program uses a birth and death process to model gene gain and loss across a user-specified phylogenetic tree. The distribution of family sizes generated under this model can provide a basis for assessing the significance of the observed family size differences among taxa. In breaf, CAFE try to estimates ancestral states of gene families comparing the tip states (one continouse state for each gene family).

---

<br/>

## PREPARING THE DATA

Preparing the data for CAFE it's a really tedious process. Fortunatly Orthofinder already provide us the Orthogroup table which is almost in the correct format. We need just a little bit of bash scripting to change few things:

```
mkdir Analyses/CAFE
cp Analyses/Results_Mar30/Orthogroups/Orthogroups.GeneCount.tsv Analyses/CAFE/
cd Analyses/CAFE/
sed -i.old 's/_2//g' Orthogroups.GeneCount.tsv
sed -i.old 's/_//g' Orthogroups.GeneCount.tsv
cat Orthogroups.GeneCount.tsv | awk '{print tolower($0)}' > Orthogroups.GeneCount_2.tsv 
sed -i.old "s/^/NONE        /g" Orthogroups.GeneCount_2.tsv #macOS users have to put a litteral tab inside sed, Linux can just use \t
rev Orthogroups.GeneCount_2.tsv | cut -d$'\t' -f 2- | rev > Orthogroups.GeneCount_ForCafe.tsv
```

Now we need to change a bit also our time tree. First of all we need to convert the nexus file into newick. For this operation I usally use figtree. Just open the nexus treefile - ensuring to have divergence time annotated - and save it as newick. Then :

```
sed -i.old 's/_//g' New_DivergenceTime.timetree_cafe.nwk
cat New_DivergenceTime.timetree_cafe.nwk | awk '{print tolower($0)}' > New_DivergenceTime.timetree_cafe_final.nwk
```

Now we are ready to use CAFE

#### This is just an example of what we can do once we have a pylogentic hypotesis about the species under study. A correct placement of lineages and species in the tree of life is therefore not only usefull *per se*, but it's almost always necessary when we want to carry one **any** type of evolutionary and comparative study!
