At this step all of you should have a fasta file for each 1-to-1 orthologs. If this is not the case just take it from ```Analyses/Orthofinder_Results_Mar30.tar.gz``` after the decompression of the folder.

# Multiple Sequence Alignment

After we have get our strictly orthologs sequences, the next step is to infer the orthology relathionships between **each site** of each sequences. In other terms, the differences among the nucleotides/amminoacide should result **only from changes that convey descent information**.This process is called **Multiple Sequence Alignment (MSA)** and is often followed by an additional step of detection and exclusion of those alignment regions whose homology we are uncertain of. Due to the large amount of data which we process nowadays, this step is often overlooked and as a result it's quite easy to find misaligned loci in modern datasets. Moreover, from the bioinformatic poin of view, there is a lot under the hood of the aligners (see [here](https://www.sciencedirect.com/science/article/pii/S0888754317300551) for a nice review on underlying computational methods).


In this tutorial, we will use the most popular tools for multiple sequence alignment:

  * MAFFT (Original publication: [Katoh et al., 2002](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC135756/))
  * T-Coffee (Original publication: [Notredame et al., 2000](https://www.sciencedirect.com/science/article/abs/pii/S0022283600940427))
  * And subsequently filter the alignments using  the old but gold:
  * Gblocks ([Talavera & Castresana, 2007](https://academic.oup.com/mbe/article/17/4/540/1127654))
 
 Let's create a new directory inside Analyses:
 
 ```
 mkdir Analyses/1-to-1.Alignments
 ```
 
 Print the help of mafft to get some few, but usefull informations:
 
 ```
 mafft --help
 ------------------------------------------------------------------------------
  MAFFT v7.475 (2020/Nov/23)
  https://mafft.cbrc.jp/alignment/software/
  MBE 30:772-780 (2013), NAR 30:3059-3066 (2002)
------------------------------------------------------------------------------
High speed:
  % mafft in > out
  % mafft --retree 1 in > out (fast)

High accuracy (for <~200 sequences x <~2,000 aa/nt):
  % mafft --maxiterate 1000 --localpair  in > out (% linsi in > out is also ok)
  % mafft --maxiterate 1000 --genafpair  in > out (% einsi in > out)
  % mafft --maxiterate 1000 --globalpair in > out (% ginsi in > out)

If unsure which option to use:
  % mafft --auto in > out

--op # :         Gap opening penalty, default: 1.53
--ep # :         Offset (works like gap extension penalty), default: 0.0
--maxiterate # : Maximum number of iterative refinement, default: 0
--clustalout :   Output: clustal format, default: fasta
--reorder :      Outorder: aligned, default: input order
--quiet :        Do not report progress
--thread # :     Number of threads (if unsure, --thread -1)
--dash :         Add structural information (Rozewicki et al, submitted)
 ```
 
For more informations about the alghoritms see the [mafft website](https://mafft.cbrc.jp/alignment/software/algorithms/algorithms.html). Just few notes:

 * **E-INS-i** strategy is designed to work for proteins which have several conserved motifs embedded in long unalignable regions (*e.g.* fast-evolving multi-domains proteins, Fig. 1A).
 * **L-INS-i** is suitable for single-domain proteins with unalignable regions (Fig. 1B).
 * **G-INS-i** is designed to maximase alignment accuracy between full alignable regions (*e.g* domains alignments, Fig. 1C).

![Examples](https://github.com/for-giobbe/phy/blob/master/2021/Images/Mafft-1.png)
