# Multiple Sequence Alignment

After we have get our strictly orthologs sequences, the next step is to infer the orthology relathionships between **each site** of each sequences. In other terms, the differences among the nucleotides/amminoacide should result **only from changes that convey descent information**.This process is called **Multiple Sequence Alignment (MSA)** and is often followed by an additional step of detection and exclusion of those alignment regions whose homology we are uncertain of. Due to the large amount of data which we process nowadays, this step is often overlooked and as a result it's quite easy to find misaligned loci in modern datasets. Moreover, from the bioinformatic poin of view, there is a lot under the hood of the aligners (see [here](https://www.sciencedirect.com/science/article/pii/S0888754317300551) for a nice review on underlying computational methods).


In this tutorial, we will use the most popular tools for multiple sequence alignment:

  * MAFFT (Original publication: [Katoh et al., 2002](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC135756/))
  * T-Coffee (Original publication: [Notredame et al., 2000](https://www.sciencedirect.com/science/article/abs/pii/S0022283600940427))
  * And subsequently filter the alignments using  the old but gold:
  * Gblocks ([Talavera & Castresana, 2007](https://academic.oup.com/mbe/article/17/4/540/1127654))
