### INTRO:

Currently, two divergent systematic methods are commonly applied for inferring species trees: the supermatrix approach (concatenation) and the coalescent approach (gene trees are calculated and then reconciled in a species tree). You will find a couple of interesting papers on the topic at the end of this tutorial, but here we will focus on the supermatrix approach.

Before starting you should have all aligned & filtered MSA, or if you missed the previous lesson you can grab them from [here](https://github.com/for-giobbe/phy/blob/master/2021/Analyses/Precomputed_1-to-1_Alignments.tar.gz)

## CONCATENATION OF MSAs

After having aligned and filtere our genes, we will concatenate them using AMAS, a nice Python toolkit. However, as preliminar steps we need to: 
  * Convert the gblock format into standard fasta. 
  * Change all fasta headers leaving only the species name. This step is mandatory beacuse concatenation is **always** made looking at the headers.
  * Create a new working directory

```
makdir mkdir Analyses/IQ-TREE
for i in Analyses/1-to-1_Alignments/gblock/*gb; do varOG=$( echo "$i" | cut -d"." -f1); cat "$i" | tr -d " " | sed 's/_1//g' > Analyses/1-to-1_Alignments/gblock/"$varOG".t_cofee-gb.fa; rm Analyses/1-to-1_Alignments/gblock/"$i"; done
for i in Analyses/1-to-1_Alignments/gblock/*.fa; do sed -i.old 's/|.*$//g' "$i"; rm Analyses/1-to-1_Alignments/gblock/*.old; done ## "-i" and "rm" is necessary only for macOS users
```

Now let's have a look at the AMAS help:

```
AMAS.py --help
usage: AMAS <command> [<args>]

The AMAS commands are:
  concat      Concatenate input alignments
  convert     Convert to other file format
  replicate   Create replicate data sets for phylogenetic jackknife
  split       Split alignment according to a partitions file
  summary     Write alignment summary
  remove      Remove taxa from alignment
  translate   Translate DNA alignment into protein alignment
  trim        Remove columns from alignment

Use AMAS <command> -h for help with arguments of the command of interest

positional arguments:
  command     Subcommand to run

optional arguments:
  -h, --help  show this help message and exit
```

and specifically, to the ```concat``` function:

```
AMAS.py concat --help
usage: AMAS.py [-h] [-p CONCAT_PART] [-t CONCAT_OUT] [-u {fasta,phylip,nexus,phylip-int,nexus-int}] [-y {nexus,raxml,unspecified}] [-e] [-c CORES] -i IN_FILES [IN_FILES ...] -f
               {fasta,phylip,nexus,phylip-int,nexus-int} -d {aa,dna}

Concatenate input alignments

optional arguments:
  -h, --help            show this help message and exit
  -p CONCAT_PART, --concat-part CONCAT_PART
                        File name for the concatenated alignment partitions. Default: 'partitions.txt'
  -t CONCAT_OUT, --concat-out CONCAT_OUT
                        File name for the concatenated alignment. Default: 'concatenated.out'
  -u {fasta,phylip,nexus,phylip-int,nexus-int}, --out-format {fasta,phylip,nexus,phylip-int,nexus-int}
                        File format for the output alignment. Default: fasta
  -y {nexus,raxml,unspecified}, --part-format {nexus,raxml,unspecified}
                        Format of the partitions file. Default: 'unspecified'
  -e, --check-align     Check if input sequences are aligned. Default: no check
  -c CORES, --cores CORES
                        Number of cores used. Default: 1

required arguments:
  -i IN_FILES [IN_FILES ...], --in-files IN_FILES [IN_FILES ...]
                        Alignment files to be taken as input. You can specify multiple files using wildcards (e.g. --in-files *fasta)
  -f {fasta,phylip,nexus,phylip-int,nexus-int}, --in-format {fasta,phylip,nexus,phylip-int,nexus-int}
                        The format of input alignment
  -d {aa,dna}, --data-type {aa,dna}
                        Type of data
```

Now we can concatenate our alignments:

```
AMAS.py concat -i Analyses/1-to-1_Alignments/gblock/*.fa -f fasta -d aa --part-format nexus
mv concatenated.out Analyses/IQ-TREE/My_Concat.fa
mv partitions.txt Analyses/IQ-TREE/
```

As a results you should have two files, a merged MSA and a partition file, where are stored the coordinate of the gene boundaries. Let's have a look at it:

```
head -n 20 partitions.txt 
#NEXUS

Begin sets;
	charset p1_OG0000030 = 1-197;
	charset p2_OG0000031 = 198-348;
	charset p3_OG0000032 = 349-1070;
	charset p4_OG0000033 = 1071-1512;
	charset p5_OG0000034 = 1513-1925;
	charset p6_OG0000035 = 1926-2300;
	charset p7_OG0000036 = 2301-2463;
	charset p8_OG0000037 = 2464-2867;
	charset p9_OG0000038 = 2868-3281;
	charset p10_OG0000039 = 3282-3703;
	charset p11_OG0000040 = 3704-4054;
	charset p12_OG0000041 = 4055-5124;
	charset p13_OG0000042 = 5125-5285;
	charset p14_OG0000043 = 5286-5482;
	charset p15_OG0000044 = 5483-6214;
	charset p16_OG0000045 = 6215-7227;
	charset p17_OG0000046 = 7228-7383;
```
>NOTE: We are working with proteins so our partitions file just store the starting and the ending position of each gene. However, if we were using nucleotide sequences of PCGs, we might also want the coordinates of all the first, second and third codon positions. Indeed, they evolve differently due to the gen code degeneracy and codon usage bias.

Now we are ready to carry out our model selection and tree inference in a Maximum Likelihood framework...

## MODEL SELECTION AND MAXIMUM LIKELIHOOD INFERENCE

For all ML analyses we will use IQ-TREE. Some of the greates advantages of this software are:

 * It can do almost everything (take a look at the [manual](http://www.iqtree.org/doc/)).
 * Each analysis can be highly customizable.
 * Really faster compared to other softwares.
 * Very easy to run as you want (if you study a bit the manual).

For a *quick-and-dirty* Model Selection on our concat we can just type:

```
iqtree -s Analyses/IQ-TREE/My_Concat.fa -m TESTONLY
```
The ```-m TESTONLY``` word stands for  standard model selection, which tells IQ-TREE to perform ModelFinder without taking into consideration FreeRate model (due to computational limitis, look at them [here](https://www.nature.com/articles/nmeth.4285). You should just add ```-m MF ``` flag to compute a full Model Selection): this tool computes the log-likelihoods of an initial parsimony tree for many different models and the Akaike information criterion (AIC), corrected Akaike information criterion (AICc), and the Bayesian information criterion (BIC). Then ModelFinder chooses the model that minimizes the BIC score (you can also change to AIC or AICc by adding the option ```-AIC``` or ```-AICc```, respectively)

The output are:

 * **My_Concat.fa.ckp.gz:** checkpoint file.
 * **My_Concat.fa.iqtree:** summary human readble file.
 * **My_Concat.fa.log:** log file.
 * **My_Concat.fa.model.gz** computer readble result file.
 * **My_Concat.fa.treefile** Tree used for ModelFinder.

Let's have a look at the iqtree summary file, where we can find a lot of usefull informations:

```
cat Analyses/IQ-TREE/My_Concat.fa.iqtree
```

>The -m flag can also specify a model name to use during the analyses, which can be a priori specified by the user (here's a [list](http://www.iqtree.org/doc/Substitution-Models) of models implemented in ModelFinder). Usually this is done to save computational time when the best-fit model has been already pre-computed.

What we've seen until now is the process through which we select the "best" model of evolution for our sequence data, according to a metric of choice. However,
when using concatenated alignments we don't want to loose *a priori* the information of the single genes boundaries. So let's use our partition file:

```
iqtree -s Analyses/IQ-TREE/My_Concat.fa -sp Analyses/IQ-TREE/partitions.txt -m TESTONLY
```
where the ```-sp``` option allows to take into consideration **[heterotachy](https://en.wikipedia.org/wiki/Heterotachy)** ( allowing each partitions to have its own set of branch lengths. **NB** very parameter rich, take a look at [this](https://academic.oup.com/mbe/article/37/4/1202/5673393?login=true) pubblication). Other options are:

* ```-q```:   all partitions share the same set of branch lengths (unrealistic).
* ```-spp```: allows each partition to have its own evolution rate (recommended for typical analysis).

The resulting file should look something like this:

```
cat Analyses/IQ-TREE/partitions.txt.best_scheme.nex
#nexus
begin sets;
  charset p1_OG0000030 = 1-197;
  charset p2_OG0000031 = 198-348;
  charset p3_OG0000032 = 349-1070;
  charset p4_OG0000033 = 1071-1512;
  charset p5_OG0000034 = 1513-1925;
  charset p6_OG0000035 = 1926-2300;
  charset p7_OG0000036 = 2301-2463;
  charset p8_OG0000037 = 2464-2867;
  charset p9_OG0000038 = 2868-3281;
  charset p10_OG0000039 = 3282-3703;
  charset p11_OG0000040 = 3704-4054;
  ...
  charpartition mymodels =
    LG+G4: p8_OG0000037,
    LG+G4: p9_OG0000038,
    LG+G4: p10_OG0000039,
    LG+G4: p11_OG0000040,
    JTT+F+G4: p12_OG0000041,
    LG+G4: p13_OG0000042,
    JTT+I: p14_OG0000043,
    LG+F+G4: p15_OG0000044,
    LG+F+G4: p16_OG0000045,
    LG+G4: p17_OG0000046,
    LG+G4: p18_OG0000047,
    ...
end; 
```
Here we can see the best-fit model for each partition, but take a look at the iqtree file for detailed informations.
Now, we can rerun the analyses using the ```-spp``` option:

```
iqtree -s Analyses/IQ-TREE/My_Concat.fa -spp Analyses/IQ-TREE/partitions.txt -m TESTONLY --prefix spp
```
The choice between -spp and -sp model can be made by looking at the **BIC scores** of the resulting models

```
varSP=$( grep "Bayesian information criterion (BIC) score:"  Analyses/IQ-TREE/partitions.txt.iqtree | cut -d" " -f6)
varSPP=$( grep "Bayesian information criterion (BIC) score:" spp.iqtree | cut -d" " -f6)
echo -e "$varSP\n$varSPP"
481216.3913
478647.5903
```
As we can see in this case the best-fit partitioning scheme was found with the ```-spp option``` (the **lower** the BIC score, the **higher** the fitness of the model).

---

The previous analysis will result in separate models for each partion. Nonetheless, there are several reasons for which we wanto to **merge partitions** which can be described by similar models of evolution, possibly including a better estimation of model parameters (less parameters to estimates = better estimation). ModelFinder implements a *greedy* strategy ([Lanfear et al., 2012](https://academic.oup.com/mbe/article/29/6/1695/1000514)) that starts with the full partition model and subsequentially merges two genes until the model fit does not increase any further. Since this process can be very long when analyszing a lot of partitions, ModelFinder implements the *relaxed hierarchical clustering algorithm* ([Lanfear et al., 2014](https://bmcecolevol.biomedcentral.com/articles/10.1186/1471-2148-14-82)), which is invoked via ```-rcluster``` option and can save a great amount of computational time.

To carry out simultaneously model of evolution & partitioning scheme selection let's use:

```
iqtree -s Analyses/IQ-TREE/My_Concat.fa -spp Analyses/IQ-TREE/partitions.txt -m TESTMERGEONLY --prefix Analyses/IQ-TREE/spp_MergedPartitions -rcluster 10
```

where with ```-rcluster 10``` we are only examining the top 10% partition merging schemes (better to carry on a full *greedy* search in real scenarios).

Now we can have a look at the resulting partitioning scheme and BIC score:

```
cat Analyses/IQ-TREE/spp_MergedPartitions.best_scheme.nex
#nexus
begin sets;
  charset p1_OG0000030_p11_OG0000040_p41_OG0000070 = 1-197  3704-4054  20714-21356;
  charset p2_OG0000031_p3_OG0000032_p7_OG0000036_p9_OG0000038_p13_OG0000042_p16_OG0000045_p22_OG0000051_p26_OG0000055_p32_OG0000061_p33_OG0000062_p48_OG0000077_p49_OG0000078 = 198-348  349-1070  2301-2463  2868-3281  5125-5285  6215-7227  9914-10832  11579-11838  17913-18242  18243-18436  24009-24208  24209-24546;
  charset p4_OG0000033_p10_OG0000039_p12_OG0000041_p34_OG0000063 = 1071-1512  3282-3703  4055-5124  18437-18757;
  charset p5_OG0000034_p15_OG0000044_p17_OG0000046_p18_OG0000047_p21_OG0000050_p23_OG0000052_p25_OG0000054_p27_OG0000056_p35_OG0000064_p36_OG0000065_p38_OG0000067_p39_OG0000068_p43_OG0000072 = 1513-1925  5483-6214  7228-7383  7384-7908  9495-9913  10833-10985  11444-11578  11839-12219  18758-18953  18954-19237  19608-19890  19891-20275  21798-22249;
  charset p6_OG0000035_p30_OG0000059 = 1926-2300  17123-17421;
  charset p8_OG0000037_p19_OG0000048_p42_OG0000071_p47_OG0000076 = 2464-2867  7909-9018  21357-21797  23553-24008;
  charset p14_OG0000043_p31_OG0000060_p46_OG0000075 = 5286-5482  17422-17912  23402-23552;
  charset p20_OG0000049_p45_OG0000074 = 9019-9494  23002-23401;
  charset p24_OG0000053 = 10986-11443;
  charset p28_OG0000057_p29_OG0000058 = 12220-16792  16793-17122;
  charset p37_OG0000066_p40_OG0000069_p44_OG0000073 = 19238-19607  20276-20713  22250-23001;
  charpartition mymodels =
    LG+I+G4: p1_OG0000030_p11_OG0000040_p41_OG0000070,
    LG+F+G4: p2_OG0000031_p3_OG0000032_p7_OG0000036_p9_OG0000038_p13_OG0000042_p16_OG0000045_p22_OG0000051_p26_OG0000055_p32_OG0000061_p33_OG0000062_p48_OG0000077_p49_OG0000078,
    LG+G4: p4_OG0000033_p10_OG0000039_p12_OG0000041_p34_OG0000063,
    LG+F+I+G4: p5_OG0000034_p15_OG0000044_p17_OG0000046_p18_OG0000047_p21_OG0000050_p23_OG0000052_p25_OG0000054_p27_OG0000056_p35_OG0000064_p36_OG0000065_p38_OG0000067_p39_OG0000068_p43_OG0000072,
    LG+G4: p6_OG0000035_p30_OG0000059,
    LG+G4: p8_OG0000037_p19_OG0000048_p42_OG0000071_p47_OG0000076,
    LG+G4: p14_OG0000043_p31_OG0000060_p46_OG0000075,
    LG+I+G4: p20_OG0000049_p45_OG0000074,
    LG+F+G4: p24_OG0000053,
    LG+I+G4: p28_OG0000057_p29_OG0000058,
    LG+G4: p37_OG0000066_p40_OG0000069_p44_OG0000073;
end;

varMerged=$(grep "Bayesian information criterion (BIC) score:" Analyses/IQ-TREE/spp_MergedPartitions.iqtree | cut -d" " -f6
echo -e "$varSP\n$varSPP\n$varMerged"
481216.3913
478647.5903
477711.6257
```


