## Estimate divergence time in a Maximum Likelihood framework

Traditionally, divergence time estimation was performed through **Bayesian analyses** (*e.g* Beast) and this is still the common and prefered practice when analyzing from few to a moderant amount of loci. However nowadays, the large amount avaible data from NGS project has made the use of software based on maximum likelihood more and more frequent, especially when the divergence time estimation is not the main goal of the research (in that case Beast is still more used). Bayesian analyses is indeed able to naturally take into account and display a lot of different sources of uncertainty, from model selection (actually it is not a model selection but a  **model averanging**), to topology and obviously divergence time. Moreover Bayesian inference has the great advantage (but also disadvantage) to be able to incorporate prior in the form of probability distributions (we will see a lit bit more in the next step). However, it is also a double edge sword, since the long computational time this type of analysis requires.

As previusly said ML can overcome this issues and, as almost everything, IQ-TREE is able to do also this! Indedd it implements the least square dating (LSD2) method to build a time tree when you have date information for tips or ancestral nodes ([here](https://academic.oup.com/sysbio/article/65/1/82/2461506) the link for the original paper).

This are the mains functions provided by IQ-TREE:

![](https://github.com/for-giobbe/phy/blob/master/2021/Images/LSD.png)

**Tip dating** is commonly used in two different scenarios:

 * 1. When analyzing viruses
 * 2. When your have a tree builded up including extinct taxon.

**Ancestral dates** are used when you have informations from fossils and you want to constrain, usually between two boundaries, the age of a/multiple specific node/s.

In this tutorial we are going only towoard this latest way.

---

As you may remember from the last part, our ML tree was not completely in agreement with the state of the art of Mollusck phylogeny. If we don't want to carry on a constrained ML tree search, this must be taken into consideration when we will choose the calibration points of our tree. Moreover, it's a good practice to try to balance calibration points, choosing both deep ones (in the backbone), and more shallow ones...

**But how can we chose which nodes to calibrates and how to calibrate them?**

As previusly said we can use data from fossils (remember that the presence of a fossil at a certain age implies that that lineage had already diversified at that age ). A very nice and comprehensive fossil database can be found [here](http://fossilworks.org/bridge.pl). However, the choice of fossils and calibration points is not so obvious and can upset the estimate of the divergence times (as well as being questioned by peers). So, a more straightforward way is to just use the same calibration scheme used in previous studies or even the already computed divergence time of certain nodes. In this tutorial we are going for the more straightforward way, taking again the [Kocot et al., 2020](https://www.nature.com/articles/s41598-019-56728-w) pubblication as a reference (the author is one of the major expert in Mollusk phylogeny).

>**NB** The way we choose to carry on the calibration depends a lot on the main aim of your research. If your main goal is the estimation of the divergence time in itself, probably is better to spend some (a lot) of time in studying the fossil and the litterature or rely on more experienced collaborators. If we just need a reliable time tree for further analyses we can chose the more straigthforward way.

Let'shave a look at some of their dicvergence time results (obtained using Beast and present in the suppl. informations of the article):

| Node                              | Divergence time (Mya)    | 95% HPD (Mya)     |
| ----------------------------------|:------------------------:| -----------------:|
| Mollusca                          | 545.449                  | 540.386 - 552.405 |
| Monoplacophora                    | 276.956                  | 85.269 - 498.414. |
| Bivalvia + Gastropoda + Scaphopoda| 529.765                  | 525.529 - 534.738 |
| Gastropoda                        | 424.426                  | 408.354 - 439.456 |
| Ganglionata                       | 533.595                  | 527.025 - 540.918 |

Where 95% HPD = Confidence interval of the estimation
Ganglionata = Gastropoda + Scaphopoda + Cephalopoda + Bivalvia
Node = Start of divergence of the relative clade

 * Now look at our ML tree previously computed and compared it to their, which branches do you think they have **in common**? Which of these calibration points can we therefore use?

<img src="https://github.com/for-giobbe/phy/blob/master/2021/Images/Kocot_et.al_2020.png" width="400">

> Remember that we have only **2x Gasteropoda**, **3x Bivalvia**, **1x Cephalopoda**, **1x Poliplacophora**, **1x Anellidae** (as outgroup). See [here](https://github.com/for-giobbe/phy/blob/master/2021/0.WorkingDirectory_Dataset.md) for taxonomy of our samples. Whatch out aslo for differences in taxon sampling!

In my opinion we can only confidentially set 3 calibration points: (1) the divergence between **Bivalvia + Gastropoda + Scaphopoda**, (2) the divergence of **Mollusks**, (3) the divergence of **Gastropoda**. Indeed: 
  
   * **1.** We don't have Monoplacophora samples (not avaible genomes).
   * **2.** In our tree we don't have the Ganglionata clade (since *A. granulata* fall between cephalopds and Gasteropoda + Mollusca)

>**NB** A quick note about the calibration of the Gasteropoda node: Here we are calibrating the crown of Gasteropoda, means when they start to diverge. We can use their estimation **ONLY** beacuse we have both *Lotita* and *Biomphalaria*, which is in a strict relathionship with *Aplysia*. So, the two species belong to the two different clades of Gasteropoda recovered by the authors. In other terms, even if the taxon sampling is different the node is the same!

Here a home made figure that could explain it better than words.

<img src="https://github.com/for-giobbe/phy/blob/master/2021/Images/Common_nodes-1.png" width="600">

