# GENE FAMILIES EVOLUTIONARY ANALYSES

## INTRO: 

Evolutionary studies about gene families are becoming very frequent in recent years. For example, in almost all genome paper is present a gene families expansion/contractions analyses. Indeed, has been observed how enormous amount of changes in families size can take place even between closely related organisms. There is much interest in these changes, as even the gain or loss of single genes have been implicated in adaptive divergence between species. In addition, large contractions and expansions of gene families are generally attributed to natural selection.

One of the most popular software is **[CAFE](https://academic.oup.com/bioinformatics/article/22/10/1269/237347)**, even if other apporaches based on gene and species tree exist. Its purpose is to analyze changes in gene family size in a way that accounts for phylogenetic history and provides a statistical foundation for evolutionary inferences. The program uses a birth and death process to model gene gain and loss across a user-specified phylogenetic tree. The distribution of family sizes generated under this model can provide a basis for assessing the significance of the observed family size differences among taxa. In breaf, CAFE try to estimates ancestral states of gene families comparing the tip states (one continouse state for each gene family).

In this tutorial we are going to use the version 4, since it is easier to install, but recently has been relased the Beta version 5, which includes more complex model of evolution such as gamma models.

---

<br/>

## PREPARING THE DATA

Preparing the data for CAFE it's a really tedious process. Fortunatly Orthofinder already provide us the Orthogroup table which is almost in the correct format. We need just a little bit of bash scripting to change few things:

```
mkdir Analyses/CAFE
cp Analyses/Results_Mar30/Orthogroups/Orthogroups.GeneCount.tsv Analyses/CAFE/
cd Analyses/CAFE/
sed -i.old 's/_2//g' Orthogroups.GeneCount.tsv
sed -i.old 's/_//g' Orthogroups.GeneCount.tsv
cat Orthogroups.GeneCount.tsv | awk '{print tolower($0)}' > Orthogroups.GeneCount_2.tsv 
sed -i.old "s/^/NONE        /g" Orthogroups.GeneCount_2.tsv #macOS users have to put a litteral tab inside sed, Linux can just use \t
rev Orthogroups.GeneCount_2.tsv | cut -d$'\t' -f 2- | rev > Orthogroups.GeneCount_ForCafe.tsv
```

Now we need to change a bit also our time tree. First of all we need to convert the nexus file into newick. For this operation I usally use figtree. Just open the nexus treefile - ensuring to have divergence time annotated - and save it as newick (if you don't have figtree installed just take mine [here](https://github.com/jacopoM28/phy/blob/master/2021/Data/Fixed_DivergenceTime.timetree.nwk). Then :

```
sed -i.old 's/_//g' New_DivergenceTime.timetree_cafe.nwk
cat New_DivergenceTime.timetree_cafe.nwk | awk '{print tolower($0)}' > New_DivergenceTime.timetree_cafe_final.nwk
```

Lastly we need to print on the screen our modified tree and the files that are present in our CAFE working directory (you will see why...)

```
ll
cat New_DivergenceTime.timetree_cafe_final.nwk
```

Now we are ready! Let's open the CAFE shell :

```
/Users/jacopomartelossi/Desktop/phylogenetic_software/CAFE/release/cafe
#
```

If the hastags appears, it means that we are inside the cafe interface. Now we need to run the following commands (the ones after the hastag) :

```
(base) Air-di-jacopo:CAFE jacopomartelossi$ /Users/jacopomartelossi/Desktop/phylogenetic_software/CAFE/release/cafe 
# #!cafe
# tree (((agranulata:484.23,((bglabrata:427.639,lgigantea:427.639):55.2575,((cvirginica:309.058,pfucata:309.058):150.185,sconstricta:459.242):23.6545):1.33305):61.2192,obimaculoides:545.449):428.147,hrobusta:973.596)
(((agranulata:484.23,((bglabrata:427.639,lgigantea:427.639):55.2575,((cvirginica:309.058,pfucata:309.058):150.185,sconstricta:459.242):23.6545):1.33305):61.2192,obimaculoides:545.449):428.147,hrobusta:973.596)
# load -i Orthogroups.GeneCount_ForCafe.tsv --filter -l log.txt
-----------------------------------------------------------
Family information: Orthogroups.GeneCount_ForCafe.tsv
Log: log.txt
Tree: (((agranulata:484.23,((bglabrata:427.639,lgigantea:427.639):55.2575,((cvirginica:309.058,pfucata:309.058):150.185,sconstricta:459.242):23.6545):1.33305):61.2192,obimaculoides:545.449):428.147,hrobusta:973.596)
The number of families is 101
Root Family size : 1 ~ 219
Family size : 0 ~ 225
P-value: 0.01
Num of Threads: 1
Num of Random: 1000
#lambda -s
Empirical Prior Estimation Result: (15 iterations)
Poisson lambda: 0.540076 & Score: inf
Lambda : 0.00006176118681 & Score: -8545.676708
.Lambda : 0.00006484924615 & Score: -8434.070271
.Lambda : 0.00006793730549 & Score: -8327.892760
.Lambda : 0.00007102536483 & Score: -8226.643199
.Lambda : 0.00007720148351 & Score: -8037.378505
.Lambda : 0.00008337760219 & Score: -7863.494925
.Lambda : 0.00009572983955 & Score: -7553.326549
.Lambda : 0.00010808207691 & Score: -7283.236002
.Lambda : 0.00013278655163 & Score: -6831.242314
.Lambda : 0.00015749102635 & Score: -6463.274760
.Lambda : 0.00020689997580 & Score: -5889.909255
.Lambda : 0.00025630892524 & Score: -5455.764681
.Lambda : 0.00035512682413 & Score: -4827.497334
.Lambda : 0.00045394472302 & Score: -4386.125952
.Lambda : 0.00065158052080 & Score: -3796.960956
.Lambda : 0.00084921631858 & Score: -3418.705641
.WARNING: Calculated posterior probability for family og0000000 = 0
Lambda : 0.00124448791413 & Score: -inf
.WARNING: Calculated posterior probability for family og0000000 = 0
Lambda : 0.00104685211636 & Score: -inf
.Lambda : 0.00065158052080 & Score: -3796.960956
.WARNING: Calculated posterior probability for family og0000000 = 0
[...]
#report report
Running Viterbi algorithm....
Report Done
# 
```

Note that the tree must be passed manually copying and pasting the newick file. Moreover, there is no selfcomplete inside cafe interface.
With the ```tree``` line we are pecifing on which tree cafe will work; with the ```load``` line which file has to use for gene families count. Here we have added the ```--filter``` to filter Orthogroups that were not present at the root - based on a parsimony approach - in order to improve parameter estimation. With the 
```lambda -s``` option we are telling cafe to estimate one birth-death model for the whole tree (we can also choose to use different lambda values for different part of the trees and then find the best-fit model with simulations and likelihood ratio tests, see the [MANUAL](https://iu.app.box.com/v/cafetutorial-pdf).
Finally with the ```report``` command we want to summarize our results.

Now can take a look at the report.cafe file:

  * On the second line, you will find the estimated value ofλfor the whole tree.
  * The line that starts with ‘# IDs of nodes’ gives us the number that will representeach species and internal node.
  * The number that appears after a species name in the tree given under ‘Newick’ is the gene count for that species and thatgene  family.   The  third  column  (‘Family-wide P-value’)  tells  us  for  each  genefamily whether it has a significantly greater rate of evolution.  When this valueis<0.01, then the fourth column (‘Viterbi P-value’) allows the identification of which branches the shift in λ was significant. In our case we can identify only one family with a p-value < 0.01 (OG0000018).
 
Now we can further summarise our results with [this](https://github.com/jacopoM28/phy/blob/master/2021/scripts/Cafe_scripts.zip) scripts - don't move them from their directoy, just use their absolute or relative path. **IMPORTANT :** The ```cafetutorial_report_analysis.py``` scripts need python 2.7, while ```cafetutorial_draw_tree.py``` python 3. You can just enter in their respective conda environment.

```
 conda activate py_2.7
 python /Users/jacopomartelossi/Desktop/phylogenetic_software/Cafe_scripts/cafetutorial_report_analysis.py -i report.cafe -o summary
```

This script should produce a few files that will help us visualize the results.

Finnally we can roughly plot all expansion/conractions/fast-evolving families on our tree. For this not easy purpose we need to print on the screen :

  * 1. Our Time tree.
  * 2. The tree with CAFE nodes labels (3rd line of report.cafe)

**NB :** Remember to activate the phyton3 environment!

```
conda deactivate
conda activate bio_info-env
head -n report.cafe
cat Fixed_DivergenceTime.timetree_cafe_final.nwk 
```

Now we are ready:

```
jacopomartelossi$ python /Users/jacopomartelossi/Desktop/phylogenetic_software/Cafe_scripts/cafetutorial_draw_tree.py -i summary_node.txt -t '((((((cvirginica:328.737,pfucata:328.737):164.705,sconstricta:493.442):36.3234,(bglabrata:424.426,lgigantea:424.426):105.339):1.0,agranulata:530.765):14.684,obimaculoides:545.449):461.882,hrobusta:1007.33);' -d '((((((cvirginica<0>,pfucata<2>)<1>,sconstricta<4>)<3>,(bglabrata<6>,lgigantea<8>)<7>)<5>,agranulata<10>)<9>,obimaculoides<12>)<11>,hrobusta<14>)<13>' -o rapid.png -y Rapid
(base) MacBook-Air-di-jacopo:CAFE jacopomartelossi$ python /Users/jacopomartelossi/Desktop/phylogenetic_software/Cafe_scripts/cafetutorial_draw_tree.py -i summary_node.txt -t '((((((cvirginica:328.737,pfucata:328.737):164.705,sconstricta:493.442):36.3234,(bglabrata:424.426,lgigantea:424.426):105.339):1.0,agranulata:530.765):14.684,obimaculoides:545.449):461.882,hrobusta:1007.33);' -d '((((((cvirginica<0>,pfucata<2>)<1>,sconstricta<4>)<3>,(bglabrata<6>,lgigantea<8>)<7>)<5>,agranulata<10>)<9>,obimaculoides<12>)<11>,hrobusta<14>)<13>' -o rapid.png -y Rapid
```

With the ```-y``` option we are specifing to print **ONLY** fast-evolcing families. But we can also plot all expansion and contractions using the ```-y Expansion``` and ```-y Contractions``` (Remember to change the output file).

After we have identified fast evolving gene families we can functionally annotate them with for example GO terms, Blast against Uniprot/SwissProt or at the sub-protein level using Interproscan. If we are luckly enough, we may found some "signal" that we can link to specific "characters" (*e.g* switch in feeding behaviour, colonization of new echological niches, emergence of new phenotypes ecc...)

---

#### This is just an example of what we can do once we have a pylogentic hypotesis about the species under study. A correct placement of lineages and species in the tree of life is therefore not only usefull *per se*, but it's almost always necessary when we want to carry one **any** type of evolutionary and comparative study!
